python/cuml/CMakeLists.txt

```diff
@@ -152,22 +77,86 @@ if(CUML_CPU)
 
   list(APPEND CYTHON_FLAGS
   "--compile-time-env GPUBUILD=0")
+
+# cuml-cpu does not need libcuml++.so
 else()
+
+  include(rapids-cpm)
+  include(rapids-export)
+  rapids_cpm_init()
+
+  # --- CUDA --- #
+  set(CUDA_STATIC_RUNTIME ON)
+
+  # Link to the CUDA wheels with shared libraries for CUDA 12+
+  #
+  # This is here because we're rebuilding cuVS below...without it, cuVS on CUDA 11
+  # dynamically links to CUDA math libs (cuBLAS, cuFFT, etc.), and then
+  # that linkage results in those libraries being vendored in wheels by auditwheel.
+  find_package(CUDAToolkit REQUIRED)
+  if(CUDAToolkit_VERSION VERSION_GREATER_EQUAL 12.0)
+    set(CUDA_STATIC_MATH_LIBRARIES OFF)
+  else()
+    if(USE_CUDA_MATH_WHEELS)
+      message(FATAL_ERROR "Cannot use CUDA math wheels with CUDA < 12.0")
+    endif()
+    set(CUDA_STATIC_MATH_LIBRARIES ON)
+  endif()
+
+  # --- RAFT---#
+  # find RAFT before cuVS, to avoid
+  # cuVS CMake defining conflicting versions of targets like 'nvidia::cutlass::cutlass'
+  include(${CUML_CPP_SRC}/cmake/thirdparty/get_raft.cmake)
+
+  # --- cuVS --- #
+  # Once there are 'libcuvs' wheels, it should be possible to remove this CPM build of cuvs.
+  #
+  #   * conda builds will find 'libcuvs' in the build environment
+  #   * wheel builds will find 'libcuvs' wherever that wheel is installed
+  #
+  # Until then, this is necessary because cuVS is included in the public headers of both
+  # libcuml (C++) and cuml (Cython).
+  set(CUML_USE_CUVS_STATIC OFF)
+  set(CUML_EXCLUDE_CUVS_FROM_ALL ON)
+  include(${CUML_CPP_SRC}/cmake/thirdparty/get_cuvs.cmake)
```

Comment:
Some of the Cython code has direct uses cuVS:

https://github.com/rapidsai/cuml/blob/d95cae5fb27ba076f7b295697cf27789e28d93fa/python/cuml/cuml/cluster/kmeans_utils.pxd#L39

so cuVS is needed in the build environment for both `libcuml` and `cuml` wheels. That means we end up compiling `libcuvs.so` in every `libcuml` build AND every `cuml` build.

Even if with decent cache hit rates, on this PR I've seen that result in it taking on the order of 2.5 hours end-to-end for all the `build-libcuml` and `build-cuml` jobs to complete :/

Maybe we need to stop here and try to add a `libcuvs` wheel?

Reply:
> Maybe we need to stop here and try to add a `libcuvs` wheel?

Yes, we agreed to do that. Trying that here: https://github.com/rapidsai/cuvs/pull/594

Resolving this thread.